fedimint-core/src/core.rs

```diff
@@ -240,192 +267,99 @@ macro_rules! newtype_impl_display_passthrough_with_instance_id {
-    fn decode_input(
-        &self,
-        r: &mut dyn io::Read,
-        instance_id: ModuleInstanceId,
-    ) -> Result<DynInput, DecodeError>;
-
-    /// Decode `Output` compatible with this module, after the module key prefix
-    /// was already decoded
-    fn decode_output(
-        &self,
-        r: &mut dyn io::Read,
-        instance_id: ModuleInstanceId,
-    ) -> Result<DynOutput, DecodeError>;
+/// A type that has a `Dyn*`, type erased version of itself
+pub trait IntoDynInstance {
+    /// The type erased version of the type implementing this trait
+    type DynType;
 
-    /// Decode `OutputOutcome` compatible with this module, after the module key
-    /// prefix was already decoded
-    fn decode_output_outcome(
-        &self,
-        r: &mut dyn io::Read,
-        instance_id: ModuleInstanceId,
-    ) -> Result<DynOutputOutcome, DecodeError>;
-
-    /// Decode `ConsensusItem` compatible with this module, after the module key
-    /// prefix was already decoded
-    fn decode_consensus_item(
-        &self,
-        r: &mut dyn io::Read,
-        instance_id: ModuleInstanceId,
-    ) -> Result<DynModuleConsensusItem, DecodeError>;
+    /// Convert `self` into its type-erased equivalent
+    fn into_dyn(self, instance_id: ModuleInstanceId) -> Self::DynType;
 }
 
-// TODO: use macro again
-/// Decoder for module associated types
-#[derive(Clone)]
-pub struct DynDecoder(Arc<dyn IDecoder + Send + Sync>);
-
-impl std::ops::Deref for DynDecoder {
-    type Target = dyn IDecoder + Send + Sync + 'static;
+type DecodeFn =
+    for<'a> fn(Box<dyn Read + 'a>, ModuleInstanceId) -> Result<Box<dyn Any>, DecodeError>;
 
-    fn deref(&self) -> &<Self as std::ops::Deref>::Target {
-        &*self.0
-    }
+#[derive(Default)]
+pub struct DecoderBuilder {
+    decode_fns: BTreeMap<TypeId, DecodeFn>,
 }
 
-impl<T> From<T> for DynDecoder
-where
-    T: Decoder + Send + Sync + 'static,
-{
-    fn from(value: T) -> Self {
-        DynDecoder(Arc::new(value))
+impl DecoderBuilder {
+    pub fn build(self) -> Decoder {
+        Decoder {
+            decode_fns: Arc::new(self.decode_fns),
+        }
     }
-}
 
-impl std::fmt::Debug for DynDecoder {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        std::fmt::Debug::fmt(&self.0, f)
+    /// Attach decoder for a specific `Type`/`DynType` pair where `DynType =
+    /// <Type as IntoDynInstance>::DynType`.
+    ///
+    /// This allows calling `decode::<DynType>` on this decoder, returning a
+    /// `DynType` object which contains a `Type` object internally.
+    ///
+    /// **Caution**: One `Decoder` object should only contain decoders that
+    /// belong to the same [*module kind*](fedimint_core::core::ModuleKind).
+    ///
+    /// # Panics
+    /// * If multiple `Types` with the same `DynType` are added
+    pub fn with_decodable_type<Type>(&mut self)
+    where
+        Type: IntoDynInstance + Decodable,
+        Type::DynType: 'static,
```

Comment:
just a question: is `DynType` always `'static` (so it could be moved to the associated type definition in the trait  itself)? 
if my regex isn't lying to me, `IntoDynInstance` is currently implemented for these types:

```sh
$ (pr/1661 *%) | rg -i 'impl.*?IntoDynInstance'
fedimint-core/src/core.rs
178:        impl fedimint_core::core::IntoDynInstance for $input {
188:        impl fedimint_core::core::IntoDynInstance for $output {
198:        impl fedimint_core::core::IntoDynInstance for $outcome {
208:        impl fedimint_core::core::IntoDynInstance for $ci {
```

e.g. the `DynInput` is expanded: 

```rust
pub struct DynInput(Box<dyn IInput + 'static + Send + Sync>, ::fedimint_core::core::ModuleInstanceId);
```
so, they're always `'static`, right? 

do we might have cases in the future where  `DynType` would not be `'static`  and, therefore, would be limiting to put the `type DynType: 'static;` bound in the trait itself?

just wondering

Reply:
non-`'static` would mean that the given value holds some references to other things, which would get in a way of treating it as "opaque value", so I don't think we would ever want it.

Reply:
right, so we could move the `'static` bound into the trait `IntoDynInstance` where the associated type is defined: `type DynType: 'static`?

Reply:
Possibly yes. Might be a good idea.

Reply:
I know that it compiles because I tried it. 
I was especially interested about the reasoning. So yeah, as we treat `DynType` as an "opaque value" I think it sounds like a good idea to have in the trait definition